File: scripts/verify_vote.ts
import { Connection, PublicKey } from "@solana/web3.js";
import * as dotenv from "dotenv";

// 导入我们需要的候选人账户解码器
import { getCandidateAccountDecoder } from "../generated/ts/voting/accounts";

dotenv.config();

// --- 脚本配置 ---
const CONFIG = {
  rpcUrl: process.env.RPC_URL || "https://api.devnet.solana.com",
  // 这是我们在第二步中添加并为其投票的候选人账户地址
  candidateAccountPubkey: new PublicKey(
    "GZzVP862HEb4dW8VJ5Loixju4dnAFDkApzVbsu2jh6x5",
  ),
};

async function main() {
  console.log("--- 🚀 Starting [Step 4: Verify Vote Result] Script ---");

  try {
    const connection = new Connection(CONFIG.rpcUrl, "confirmed");

    console.log(
      `🔍 Checking candidate account: ${CONFIG.candidateAccountPubkey.toBase58()}`,
    );

    // 1. 从区块链获取账户信息
    const accountInfo = await connection.getAccountInfo(
      CONFIG.candidateAccountPubkey,
    );
    if (!accountInfo) {
      throw new Error("Candidate account not found on the blockchain.");
    }

    // 2. 使用生成的解码器来解析二进制数据
    const decodedCandidate = getCandidateAccountDecoder().decode(
      accountInfo.data,
    );

    const candidateName = decodedCandidate.name;
    const voteCount = decodedCandidate.votes;

    // 3. 打印出验证结果
    console.log("\n✅ Verification Successful!");
    console.log(`   - Candidate Name: "${candidateName}"`);
    console.log(`   - Vote Count: ${voteCount}`);

    // 4. 最终确认
    if (voteCount > 0) {
      console.log(`\n🎉🎉 Great! The vote was correctly recorded on-chain.`);
    } else {
      console.log(
        `\n🤔 Hmm, the vote count is still 0. Something might be wrong.`,
      );
    }
  } catch (error) {
    console.error("\n❌ Script failed:", error);
  }
}

main();

/*
voting on  master [!?] via ⬢ v23.11.0 via 🍞 v1.2.17 via 🦀 1.88.0 took 4.0s
➜ bun run scripts/verify_vote.ts
[dotenv@17.2.0] injecting env (0) from .env (tip: 🔐 encrypt with dotenvx: https://dotenvx.com)
--- 🚀 Starting [Step 4: Verify Vote Result] Script ---
🔍 Checking candidate account: GZzVP862HEb4dW8VJ5Loixju4dnAFDkApzVbsu2jh6x5

✅ Verification Successful!
   - Candidate Name: "Candidate #1"
   - Vote Count: 1

🎉🎉 Great! The vote was correctly recorded on-chain.
*/

File: rust-client/examples/verify_vote.rs
use anyhow::Result;
use solana_client::rpc_client::RpcClient;
use solana_sdk::pubkey::Pubkey;
use std::{env, str::FromStr};

use voting_client::accounts::CandidateAccount;

#[tokio::main]
async fn main() -> Result<()> {
    println!("--- 🚀 Starting [Step 4: Verify Vote] Rust Client ---");

    dotenvy::from_path("../.env").ok();

    let rpc_url =
        env::var("RPC_URL").unwrap_or_else(|_| "https://api.devnet.solana.com".to_string());

    // !! 重要：请将这里的地址替换为您投票的候选人地址 !!
    let candidate_account_pubkey =
        Pubkey::from_str("D2dmKitcUCCcDYEsd1vT67rzF11xTV1kDZwJoXhRnet1")?;

    let client = RpcClient::new(rpc_url);

    println!(
        "🔍 Checking candidate account: {}",
        candidate_account_pubkey
    );

    let account_info = client.get_account(&candidate_account_pubkey)?;
    let candidate_data = CandidateAccount::from_bytes(&account_info.data)?;

    println!("\n✅ Verification Successful!");
    println!("   - Candidate Name: \"{}\"", candidate_data.name);
    println!("   - Vote Count: {}", candidate_data.votes);

    if candidate_data.votes > 0 {
        println!("\n🎉🎉 Great! The vote was correctly recorded on-chain.");
    } else {
        println!("\n🤔 Hmm, the vote count is still 0. Something might be wrong.");
    }

    Ok(())
}
